File: scripts/algorithms/S/Sum of Distances/Sum of Distances..java
// Runtime: 114 ms (Top 5.75%) | Memory: 89.00 MB (Top 5.75%)

class Solution {
    public long[] distance(int[] nums) {
        long ans[]=new long[nums.length];
        int arr[][]=new int[nums.length][2];
        for(int i=0;i<nums.length;i++){
            arr[i][0]=nums[i];
            arr[i][1]=i;
        }
        Arrays.sort(arr,(a,b)->a[0]==b[0]?a[1]-b[1]:a[0]-b[0]);
        HashMap<Integer,List<Integer>> h=new HashMap<>();
         HashMap<Integer,Long> h1=new HashMap<>();
        long s=0;
        for(int i=0;i<nums.length;i++){
            if(h.containsKey(arr[i][0])){
                s+=1L*arr[i][1];
                h1.put(arr[i][1],s);
            }else{
                s=arr[i][1];
                h1.put(arr[i][1],s);
            }
             h.putIfAbsent(arr[i][0],new ArrayList<>());
            h.get(arr[i][0]).add(arr[i][1]);
        }
        for(int i=0;i<nums.length;i++){
            if(h.get(nums[i]).size()==1){
                ans[i]=0;
            }else{
                List<Integer> l1=h.get(nums[i]);
                int ind=Collections.binarySearch(l1,i);
                long leftSum=1L*(h1.get(i).longValue()-i);
                long rightSum=1L*(h1.get(l1.get(l1.size()-1)).longValue()-leftSum-i);
                long x=1L*(ind)*i-leftSum-1L*(l1.size()-ind-1)*i+rightSum;
                ans[i]=x;
            }
        }
        return ans;
    }
}

File: scripts/algorithms/S/Sum of Distances/Sum of Distances.cpp
// Runtime: 159 ms (Top 81.54%) | Memory: 109.80 MB (Top 66.77%)

class Solution {
public:
    vector<long long> distance(vector<int>& nums) {
        unordered_map<int, vector<int>> mp;
        int n = nums.size();        
        vector<long long> ans(n, 0);
        
        for(int i = 0; i < n; i++) {
            mp[nums[i]].push_back(i);
        }
        
        for(auto& itr: mp) {
            int m = itr.second.size();
            
            if(m == 1) continue;
            
            vector<long long> left(m), right(m);
            
            left[0] = 0;
            for(int i = 1; i < m; i++) {
                left[i] = left[i-1] + (long long)(itr.second[i] - itr.second[i-1]) * i;
            }
            
            right[m-1] = 0;
            for(int i = m-2; i >= 0; i--) {
                right[i] = right[i+1] + (long long)(itr.second[i+1] - itr.second[i]) * (m-i-1);
            }
            
            for(int i = 0; i < m; i++) {
                ans[itr.second[i]] = left[i] + right[i];
            }
        }
        
        return ans;
    }
};

